### PR TITLE
Add access AZ Resources from MOE

### PR DESCRIPTION
**Pull request by xanwal**
Mirroring CLI version for MS Docs in Python SDKv2.

# PR into Azure/azureml-examples

## Checklist

I have:

- [x] read and followed the contributing guidelines

## Changes

-

fixes #

### CODE REVIEW

**Comment by shohei1029** [comment]
I've already mentioned this in chat, just let me put it again for my visibility.
---
Thank you so much for your work and PR - Alexander. It is nicely organized and easy to read.

Only one thing coming to my mind is that can you create separate notebooks for system-assigned managed identity and user-assigned managed identity?

- They are separated for the CLI version of samples so we can align with them:
    [azureml-examples/deploy-managed-online-endpoint-access-resource-sai.sh at main · Azure/azureml-examples (github.com)](https://github.com/Azure/azureml-examples/blob/main/cli/deploy-managed-online-endpoint-access-resource-sai.sh)

    [azureml-examples/deploy-managed-online-endpoint-access-resource-uai.sh at main · Azure/azureml-examples (github.com)](https://github.com/Azure/azureml-examples/blob/main/cli/deploy-managed-online-endpoint-access-resource-uai.sh)

- I believe an end-to-end example for a single topic is easy for customers to try because they don't need to select cells to run.

**Review by shohei1029 — CHANGES_REQUESTED** [review]
I've already mentioned this in chat, just let me put it again for my visibility.

---
Thank you so much for your work and PR - Alexander. It is nicely organized and easy to read.

Only one thing coming to my mind is that can you create separate notebooks for system-assigned managed identity and user-assigned managed identity?

- They are separated for the CLI version of samples so we can align with them:
    [azureml-examples/deploy-managed-online-endpoint-access-resource-sai.sh at main · Azure/azureml-examples (github.com)](https://github.com/Azure/azureml-examples/blob/main/cli/deploy-managed-online-endpoint-access-resource-sai.sh)

    [azureml-examples/deploy-managed-online-endpoint-access-resource-uai.sh at main · Azure/azureml-examples (github.com)](https://github.com/Azure/azureml-examples/blob/main/cli/deploy-managed-online-endpoint-access-resource-uai.sh)

- I believe an end-to-end example for a single topic is easy for customers to try because they don't need to select cells to run.

**Comment by xanwal** [comment]
@shohei1029 I've pushed the split files and tagged the cells for use in AZ docs

File: sdk/endpoints/online/managed/online-endpoints-managed-identity-sai.ipynb [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "name" : "create-storage-8"
+   },
+   "outputs": [],
+   "source": [
+    "file_path = \"endpoints/online/managed/managed-identities/hello.txt\" \n",
```
Comment (shohei1029) [inline sdk/endpoints/online/managed/online-endpoints-managed-identity-sai.ipynb:233]: Could you place `hello.txt` in the same directory and add in the PR?
(You can just copy the file from `cli/endpoints/online/managed/managed-identities/hello.txt`)

File: sdk/endpoints/online/managed/online-endpoints-managed-identity-sai.ipynb [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+   "metadata": {
+    "name" : "define-configuration-3"
+   },
+   "outputs": [],
+   "source": [
+    "sai_deployment = ManagedOnlineDeployment(\n",
+    "    name=\"blue\", \n",
+    "    endpoint_name=sai_endpoint_name, \n",
+    "    model=Model(\n",
+    "        path=\"../../model-1/model/\"\n",
```
Comment (shohei1029) [inline sdk/endpoints/online/managed/online-endpoints-managed-identity-sai.ipynb:51]: Could you place these two notebooks into the managed-identities directory like `sdk/endpoints/online/managed/managed-identities` to ensure the relative path `../../model-1/` work properly?

File: sdk/endpoints/online/managed/online-endpoints-managed-identity-sai.ipynb [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+    "role_defs = auth_client.role_definitions.list(scope=scope)\n",
+    "role_def = next((r for r in role_defs if r.role_name == role_name))\n",
+    "\n",
+    "auth_client.role_assignments.create(\n",
+    "    scope=scope, \n",
+    "    role_assignment_name=uuid.uuid4(),\n",
+    "    parameters=RoleAssignmentProperties(\n",
+    "        role_definition_id = role_def.id, \n",
+    "        principal_id = system_principal_id\n",
+    "    )\n",
```
Comment (shohei1029) [inline sdk/endpoints/online/managed/online-endpoints-managed-identity-sai.ipynb:352]: I've faced the following error while running the code.
Did you face the same error, or do you have any ideas for this? Sorry..

```
HttpResponseError: (MalformedRoleAssignmentRequest) PASRP RoleAssignment request is malformed
Code: MalformedRoleAssignmentRequest
Message: PASRP RoleAssignment request is malformed
```

File: sdk/endpoints/online/managed/online-endpoints-managed-identity-uai.ipynb [hunk 4]
```diff
@@ -0,0 +1,563 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "name" : "test-deployment-1"
+   },
+   "outputs": [],
+   "source": [
+    "sample_data = \"endpoints/online/model-1/sample-request.json\"\n",
```
Comment (shohei1029) [inline sdk/endpoints/online/managed/online-endpoints-managed-identity-uai.ipynb:517]: This path seems to be a relative path from `cli` directory.
Could you change it to the relative path from the .ipynb?

File: sdk/endpoints/online/managed/online-endpoints-managed-identity-sai.ipynb (hunk 3 above)
Comment (xanwal) [reply sdk/endpoints/online/managed/online-endpoints-managed-identity-sai.ipynb]: Working on debugging this but has to be indirectly through the runner as I don't actually have role assignment permissions on any resource.

File: sdk/endpoints/online/managed/online-endpoints-managed-identity-sai.ipynb (hunk 2 above)
Comment (xanwal) [reply sdk/endpoints/online/managed/online-endpoints-managed-identity-sai.ipynb]: Change made

File: sdk/endpoints/online/managed/online-endpoints-managed-identity-sai.ipynb (hunk 1 above)
Comment (xanwal) [reply sdk/endpoints/online/managed/online-endpoints-managed-identity-sai.ipynb]: Added

File: sdk/endpoints/online/managed/online-endpoints-managed-identity-uai.ipynb (hunk 4 above)
Comment (xanwal) [reply sdk/endpoints/online/managed/online-endpoints-managed-identity-uai.ipynb]: Good catch, thank you.

File: sdk/endpoints/online/managed/online-endpoints-managed-identity-sai.ipynb (hunk 3 above)
Comment (shohei1029) [reply sdk/endpoints/online/managed/online-endpoints-managed-identity-sai.ipynb]: I see. I'll give you permission if I can..

**Comment by xanwal** [comment]
I've made a commit reorganizing the top two sections to enable variables to flow through without having to do excessive modification to the runners, I propose we flip "Configure Variables" and "Define Configuration" in the existing AZ Doc.

File: sdk/endpoints/online/managed/managed-identities/online-endpoints-managed-identity-uai.ipynb [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "name": "create-managed-identity-1"
+   },
+   "outputs": [],
+   "source": [
+    "from azure.identity import DefaultAzureCredential\n",
+    "from azure.mgmt.msi import ManagedServiceIdentityClient\n",
+    "from azure.mgmt.msi.models import Identity"
```
Comment (shohei1029) [inline sdk/endpoints/online/managed/managed-identities/online-endpoints-managed-identity-uai.ipynb:133]: I've faced an error when loading this module. `ModuleNotFoundError: No module named 'azure.mgmt.msi'`
Can we describe the required libraries at the top of the notebook, so customers can understand what they need to run the notebook?

**Comment by shohei1029** [comment]
> I've made a commit reorganizing the top two sections to enable variables to flow through without having to do excessive modification to the runners, I propose we flip "Configure Variables" and "Define Configuration" in the existing AZ Doc.

I agree with you. I believe defining necessary variables at first also makes it easier for customers to run the notebook.

File: .github/workflows/sdk-endpoints-online-managed-managed-identities-online-endpoints-managed-identity-sai.yml [hunk 6]
```diff
@@ -0,0 +1,58 @@
+1name: sdk-endpoints-online-managed-managed-identities-online-endpoints-managed-identity-sai
```
[the remark this answers is not in the document]
Comment (xanwal) [reply .github/workflows/sdk-endpoints-online-managed-managed-identities-online-endpoints-managed-identity-sai.yml:1]: Fixed in incoming commit.

File: sdk/endpoints/online/managed/managed-identities/online-endpoints-managed-identity-uai.ipynb (hunk 5 above)
Comment (xanwal) [reply sdk/endpoints/online/managed/managed-identities/online-endpoints-managed-identity-uai.ipynb:133]: Good idea - added a section

**Comment by xanwal** [comment]
> 

Above changes made and added a PR for the AZ doc here: https://github.com/MicrosoftDocs/azure-docs-pr/pull/210963

File: sdk/endpoints/online/managed/managed-identities/online-endpoints-managed-identity-sai.ipynb [hunk 7]
```diff
@@ -0,0 +1,588 @@
...
+   "source": [
+    "# Access Azure resources from an online endpoint with a system-assigned managed identity"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Additional packages are required for this example including \n",
+    "* azure.mgmt.msi\n",
```
Comment (shohei1029) [inline sdk/endpoints/online/managed/managed-identities/online-endpoints-managed-identity-sai.ipynb:15]: Do we need `azure.mgmt.msi` for using system-assigned managed identity?

**Comment by xanwal** [comment]
@shohei1029 The SAI/UAI notebooks have been updated to address that error you were receiving. Dan was able to acquire role creation permissions temporarily and we have tested this new method. The issue stemmed from differences in models/operations across REST API versions. I am updating the doc now.

**Comment by xanwal** [comment]
Do you have any idea how we can adapt the runner to accommodate the need for role assignment permissions?

**Comment by shohei1029** [comment]
> Do you have any idea how we can adapt the runner to accommodate the need for role assignment permissions?

Sorry, I have no ides..😥

**Comment by shohei1029** [comment]
@xanwal I've confirmed that the SAI notebook worked for me :)

However, I face an error for UAI notebook.
In `### 7.2 Create the deployment` section, when I tried to run `ml_client.online_deployments.begin_create_or_update(deployment)`, I face the following `ResourceNotReady` error.

image

Also, I can't find `"/azureml/ImageLogs/0708db04-4a2f-4f52-9d01-65aef943f68c/build.log"`.

**Comment by xanwal** [comment]
> In `### 7.2 Create the deployment` section, when I tried to run `ml_client.online_deployments.begin_create_or_update(deployment)`, I face the following `ResourceNotReady` error.

If you can look in the portal under the deployment and get the container logs that would be helpful. The missing build logs is a long-standing issue.

**Comment by shohei1029** [comment]
Oh, I see..
Here is the deployment log. It seems that I'm failing to pull the image from the workspace ACR.

```
Instance status:
SystemSetup: Succeeded
UserContainerImagePull: InProgress
ModelDownload: InProgress
UserContainerStart: Waiting

Container events:
Kind: Pod, Name: Downloading, Type: Normal, Time: 2022-09-15T07:56:49.022693Z, Message: Start downloading models
Kind: Pod, Name: Pulling, Type: Normal, Time: 2022-09-15T07:56:49.634091Z, Message: Start pulling container image
Kind: Pod, Name: Downloading, Type: Normal, Time: 2022-09-15T07:56:50.214539Z, Message: Start downloading models
Kind: Pod, Name: DownloadFailed, Type: Warning, Time: 2022-09-15T07:56:51.008386Z, Message: Model download failed, retrying. Please check model-mount log for detail.
Kind: Pod, Name: DownloadFailed, Type: Warning, Time: 2022-09-15T07:56:52.030423Z, Message: Model download failed, retrying. Please check model-mount log for detail.
Kind: Pod, Name: Pulling, Type: Normal, Time: 2022-09-15T07:56:56.129313Z, Message: Start pulling container image
Kind: Pod, Name: ImagePullFailed, Type: Warning, Time: 2022-09-15T07:57:01.05333Z, Message: Image pull failed, retrying. Please check image-fetcher log for detail.
Kind: Pod, Name: Downloading, Type: Normal, Time: 2022-09-15T07:57:07.092326Z, Message: Start downloading models
Kind: Pod, Name: DownloadFailed, Type: Warning, Time: 2022-09-15T07:57:09.071028Z, Message: Model download failed, retrying. Please check model-mount log for detail.
Kind: Pod, Name: Pulling, Type: Normal, Time: 2022-09-15T07:57:15.084402Z, Message: Start pulling container image
Kind: Pod, Name: DownloadFailed, Type: Warning, Time: 2022-09-15T07:57:19.896264Z, Message: Model download failed, retrying. Please check model-mount log for detail.
Kind: Pod, Name: ImagePullFailed, Type: Warning, Time: 2022-09-15T07:57:21.09675Z, Message: Image pull failed, retrying. Please check image-fetcher log for detail.
Kind: Pod, Name: ImagePullFailed, Type: Warning, Time: 2022-09-15T07:57:35.906174Z, Message: Image pull failed, retrying. Please check image-fetcher log for detail.
Kind: Pod, Name: Downloading, Type: Normal, Time: 2022-09-15T07:57:39.06189Z, Message: Start downloading models
Kind: Pod, Name: DownloadFailed, Type: Warning, Time: 2022-09-15T07:57:41.153033Z, Message: Model download failed, retrying. Please check model-mount log for detail.
Kind: Pod, Name: Pulling, Type: Normal, Time: 2022-09-15T07:57:54.11554Z, Message: Start pulling container image
Kind: Pod, Name: DownloadFailed, Type: Warning, Time: 2022-09-15T07:57:54.890668Z, Message: Model download failed, retrying. Please check model-mount log for detail.

Container logs:
+ '[' -z '' ']'
+ '[' '!' -S /run/containerd/containerd.sock ']'
+ ./mir-imagefetcher
{"level":"[INFO]","ts":"Sep  15 07:57:54","logger":"VMAgent.ImageFetcher","caller":"imagefetcher/image_fetcher.go:113","msg":"Use environment: AzurePublicCloud"}
{"level":"[INFO]","ts":"Sep  15 07:57:54","logger":"VMAgent.main","caller":"image-fetcher/main.go:52","msg":"InjectionEnabled=false"}
{"level":"[INFO]","ts":"Sep  15 07:57:54","logger":"VMAgent.ImageFetcher","caller":"imagefetcher/image_fetcher.go:250","msg":"Start getting image credential","imageURL":"mllab861eceff.azurecr.io/azureml/azureml_a6ee691263e17c228a7b37e0c987e32c:latest"}
{"level":"[INFO]","ts":"Sep  15 07:57:54","logger":"VMAgent.ImageFetcher","caller":"imagefetcher/image_fetcher.go:274","msg":"Found normal ACR image","imageURL":"mllab861eceff.azurecr.io/azureml/azureml_a6ee691263e17c228a7b37e0c987e32c:latest"}
{"level":"[INFO]","ts":"Sep  15 07:57:54","logger":"VMAgent.ImageFetcher","caller":"imagefetcher/image_fetcher.go:369","msg":"Request parameters","Registry":"mllab861eceff.azurecr.io","clientid":"3c976ce5-181e-4b3e-8b58-4cbc41caa62c"}
{"level":"[WARN]","ts":"Sep  15 07:57:54","logger":"VMAgent","caller":"common/http_util.go:173","msg":"Http status code of response is not 2xx","HTTP status code":500}
{"level":"[WARN]","ts":"Sep  15 07:57:56","logger":"VMAgent","caller":"common/code_util.go:44","msg":"After attempt # 1, got error: get msi token return non-success status: 500, details: getting MSI Token faild, http status code: 500  , resource: https://management.azure.com/  , clinet_id: 3c976ce5-181e-4b3e-8b58-4cbc41caa62c  , response body: InternalError - :{\nInfo: Request failure status code: 404\n\n}"}
{"level":"[WARN]","ts":"Sep  15 07:57:56","logger":"VMAgent","caller":"common/http_util.go:173","msg":"Http status code of response is not 2xx","HTTP status code":500}
{"level":"[WARN]","ts":"Sep  15 07:57:58","logger":"VMAgent","caller":"common/code_util.go:44","msg":"After attempt # 2, got error: get msi token return non-success status: 500, details: getting MSI Token faild, http status code: 500  , resource: https://management.azure.com/  , clinet_id: 3c976ce5-181e-4b3e-8b58-4cbc41caa62c  , response body: InternalError - :{\nInfo: Request failure status code: 404\n\n}"}
{"level":"[WARN]","ts":"Sep  15 07:57:58","logger":"VMAgent","caller":"common/http_util.go:173","msg":"Http status code of response is not 2xx","HTTP status code":500}
{"level":"[ERROR]","ts":"Sep  15 07:57:58","logger":"VMAgent.ImageFetcher","caller":"imagefetcher/image_fetcher.go:390","msg":"get msi token failed","error":"get msi token return non-success status: 500, details: getting MSI Token faild, http status code: 500  , resource: https://management.azure.com/  , clinet_id: 3c976ce5-181e-4b3e-8b58-4cbc41caa62c  , response body: InternalError - :{\nInfo: Request failure status code: 404\n\n}","stacktrace":"goms.io/azureml/mir/mir-vmagent/pkg/imagefetcher.(*ctrCommandExecutor).getRefreshToken\n\t/mnt/vss/_work/1/s/pkg/imagefetcher/image_fetcher.go:390\ngoms.io/azureml/mir/mir-vmagent/pkg/imagefetcher.(*ctrCommandExecutor).GetImageCredential\n\t/mnt/vss/_work/1/s/pkg/imagefetcher/image_fetcher.go:277\ngoms.io/azureml/mir/mir-vmagent/pkg/imagefetcher.(*ImageFetcher).PullAllImages\n\t/mnt/vss/_work/1/s/pkg/imagefetcher/image_fetcher.go:228\nmain.main\n\t/mnt/vss/_work/1/s/cmd/image-fetcher/main.go:59\nruntime.main\n\t/opt/hostedtoolcache/go/1.15.15/x64/src/runtime/proc.go:204"}
{"level":"[ERROR]","ts":"Sep  15 07:57:58","logger":"VMAgent.main","caller":"image-fetcher/main.go:36","msg":"Pull images failed: get msi token return non-success status: 500, details: getting MSI Token faild, http status code: 500  , resource: https://management.azure.com/  , clinet_id: 3c976ce5-181e-4b3e-8b58-4cbc41caa62c  , response body: InternalError - :{\nInfo: Request failure status code: 404\n\n}","stacktrace":"main.main.func1\n\t/mnt/vss/_work/1/s/cmd/image-fetcher/main.go:36\nmain.main\n\t/mnt/vss/_work/1/s/cmd/image-fetcher/main.go:61\nruntime.main\n\t/opt/hostedtoolcache/go/1.15.15/x64/src/runtime/proc.go:204"}
```

**Review by shohei1029 — APPROVED** [review]
I confirmed working! Thanks a lot..!

File: sdk/endpoints/online/managed/managed-identities/online-endpoints-managed-identity-uai.ipynb [hunk 8]
```diff
@@ -0,0 +1,849 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "name": "4-get-handle"
+   },
+   "outputs": [],
+   "source": [
+    "from azure.identity import DefaultAzureCredential, AzureCliCredential\n",
```
Comment (shohei1029) [inline sdk/endpoints/online/managed/managed-identities/online-endpoints-managed-identity-uai.ipynb:277]: Do we need to import this again?
`from azure.identity import DefaultAzureCredential, AzureCliCredential`

File: sdk/endpoints/online/managed/managed-identities/online-endpoints-managed-identity-uai.ipynb [hunk 9]
```diff
@@ -0,0 +1,849 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {
+    "name": "3-get-handle"
+   },
+   "outputs": [],
+   "source": [
+    "from azure.identity import DefaultAzureCredential\n",
```
Comment (shohei1029) [inline sdk/endpoints/online/managed/managed-identities/online-endpoints-managed-identity-uai.ipynb:200]: Do we need to import this again?

**Review by shohei1029 — APPROVED** [review]

File: sdk/endpoints/online/managed/managed-identities/online-endpoints-managed-identity-uai.ipynb (hunk 9 above)
Comment (xanwal) [reply sdk/endpoints/online/managed/managed-identities/online-endpoints-managed-identity-uai.ipynb]: Resolved, thanks

File: sdk/endpoints/online/managed/managed-identities/online-endpoints-managed-identity-uai.ipynb (hunk 8 above)
Comment (xanwal) [reply sdk/endpoints/online/managed/managed-identities/online-endpoints-managed-identity-uai.ipynb]: Resolved, thanks

**Review by dem108 — APPROVED** [review]
Thank you. Remaining workflows are not related to this change and they shouldn't have been triggered by this change. Approving to merge this change.

**Pull request merged**